Mathlib/FieldTheory/IsAlgClosed/Basic.lean

```diff
@@ -308,15 +308,21 @@ theorem maximalSubfieldWithHom_is_maximal :
 
 -- porting note: this was much faster in lean 3
 set_option synthInstance.maxHeartbeats 200000 in
-theorem maximalSubfieldWithHom_eq_top : (maximalSubfieldWithHom K L M).carrier = ⊤ := by
-  rw [eq_top_iff]
-  intro x _
+theorem mem_maximalSubfieldWithHom_of_splits (hL : Algebra.IsAlgebraic K L) (x : L)
+    (hsplit : Splits (algebraMap K M) (minpoly K x)) :
+    x ∈ (maximalSubfieldWithHom K L M).carrier := by
```

Comment:
This is like an infinitary version of [Polynomial.lift_of_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Field.html#Polynomial.lift_of_splits) and I think `hL : Algebra.IsAlgebraic K L` can be replace by `IsIntegral K x` (or IsAlgebraic).

Reply:
> I think `hL : Algebra.IsAlgebraic K L` can be replace by `IsIntegral K x` (or IsAlgebraic).

I think so, but then we should use `IntermediateField` instead of `Subalgebra` in the structure `SubfieldWithHom`, to exclude the non-field subalgebra such as `K[X]` between `K` and `K(X)`. I think `IntermediateField` is better since we're always talking about fields here, but I'm not sure if that change will work, I'm not tried yet.

Reply:
If it's just a subalgebra, I think you can still extend the AlgHom to its field of fractions, and then you can go on to adjoin the integral/algebraic element `x`. Will indeed make the proof (which is currently super slow to work with) more complicated though; I'll look into it later.

Reply:
Actually, we can also use [IntermediateField.algHom_mk_adjoin_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Adjoin.html#IntermediateField.algHom_mk_adjoin_splits), which is exactly the infinitary version of [Polynomial.lift_of_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Field.html#Polynomial.lift_of_splits) and used in the construction of [AlgHom.liftNormal](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Normal.html#AlgHom.liftNormal), and has a more versatile API than `maximalSubfieldWithHom`. Maybe we should aim to remove `maximalSubfieldWithHom` in favor of the `IntermediateField.Lifts` API.

Reply:
> Actually, we can also use [IntermediateField.algHom_mk_adjoin_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Adjoin.html#IntermediateField.algHom_mk_adjoin_splits), which is exactly the infinitary version of [Polynomial.lift_of_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Field.html#Polynomial.lift_of_splits), imported by this file, used in the construction of [AlgHom.liftNormal](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Normal.html#AlgHom.liftNormal), and has a more versatile API than `maximalSubfieldWithHom`. 

Yes, I think it can be used to prove `IsAlgClosed.lift` and `IsSepClosed.lift` without introducing `maximalSubfieldWithHom`.

> Maybe we should aim to remove `maximalSubfieldWithHom` in favor of the `IntermediateField.Lifts` API.
> 
> (Maybe not, it seems proofs for `maximalSubfieldWithHom` are more succinct. Maybe we should combine the best of both developments.)

Looks like that the type `IntermediateField.Lifts` is exactly the same as the type `SubfieldWithHom` but with `Subalgebra` replaced by `IntermediateField`. As I said before, I think `IntermediateField` is a better choice compared to `Subalgebra`. So I think it's fine to remove `SubfieldWithHom` completely. We can port the properties of `SubfieldWithHom` which are not mentioned for `IntermediateField.Lifts`.

As for `maximalSubfieldWithHom` and `IntermediateField.algHom_mk_adjoin_splits`, the latter merely says that there exists a map from `F(S)` to `K`, but the former says that `F(S)` is contained in `maximalSubfieldWithHom`, therefore the latter is a corollary of the former.

In fact the first line of the proof of `IntermediateField.algHom_mk_adjoin_splits`
```lean
  obtain ⟨x, hx⟩ : ∃ m : Lifts F E K, ∀ a, m ≤ a → a = m :=
    zorn_partialOrder Lifts.exists_upper_bound
```
The `x` is in fact the `maximalSubfieldWithHom`. So I think maybe we can define `maximalLift F E K` to be that element?

Reply:
Let me cc the original authors @tb65536 for the IntermediateField version [#5860](https://github.com/leanprover-community/mathlib/pull/5860), and @ChrisHughes24 / @jcommelin for the Subalgebra version [#9110](https://github.com/leanprover-community/mathlib/pull/9110). It would be possible to keep the generality of Subalgebra, and prove the maximal lift satisfies `inv_mem` to construct an IntermediateField when the source and the target are both fields, but I'm not sure how useful it would be, as both constructions currently assume the source, target and base ring are all fields. It may be straightforward to generalize to an arbitrary base ring with NoZeroSmulDivisors condition, but it seems proving [IsAlgClosed.lift](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/IsAlgClosed/Basic.html#IsAlgClosed.lift) (where a non-field source is allowed) still requires reducing to the field case.